jobs/audit_jobs.py

```diff
@@ -43,70 +45,58 @@ def run(self):
         existing_models, deleted_models = (
             self.pipeline
             | 'Get all models' >> self.job_options.model_getter()
-            | 'Partition by model.deleted' >> beam.Partition(
-                lambda model, unused_num_partitions: int(model.deleted), 2)
+            | 'Partition by model.deleted' >> (
+                beam.Partition(lambda model, _: int(model.deleted), 2))
         )
 
-        # TODO(#11475): Implement "ValidateModelRelationships" using this pcoll.
-        kind_and_existing_model_pairs = (
+        models_of_kind_by_index = (
             existing_models
-            | 'Map existing models to (kind, model) pairs' >> beam.Map(
-                lambda model: (jobs_utils.get_model_kind(model), model))
+            # NOTE: Partition returns a statically-sized list of PCollections.
+            # Creating partitions can be wasteful when there are only a few
+            # items, like in our unit tests. In exchange, we can take advantage
+            # of the high parallelizability of PCollections, which are designed
+            # for huge datasets and parallel processing.
+            #
+            # Alternatively, we could have used GroupBy. However, that returns
+            # an _iterable_ over the groups instead of a PCollection, and so it
+            # is vulnerable to out-of-memory errors.
+            #
+            # Since this job is concerned with running audits on EVERY MODEL IN
+            # STORAGE, Partition is the clear winner regardless of the overhead
+            # we'll see in unit tests.
+            | 'Split models into parallelizable PCollections' >> beam.Partition(
+                lambda m, _, kinds: kinds.index(jobs_utils.get_model_kind(m)),
+                # NOTE: Partition requires a hard-coded number of slices, it
+                # cannot be used with dynamic numbers generated in a pipeline.
+                # The KIND_BY_INDEX is a constant tuple, so it satisfies that
+                # requirement.
+                len(KIND_BY_INDEX), KIND_BY_INDEX)
         )
 
-        audits_by_kind = audit_decorators.AuditsExisting.get_audits_by_kind()
-        kind_and_audit_class_pairs = (
-            self.pipeline
-            | 'Compute the predefined (kind, audit class) pairs' >> beam.Create(
-                (kind, audit_class)
-                for kind, audit_classes in audits_by_kind.items()
-                for audit_class in audit_classes)
-        )
+        audit_error_pcolls = list(itertools.chain.from_iterable(
```

Comment:
There is a lot going on here, so perhaps it might be worth adding a comment to explain it a bit. I could follow the previous flow easily but this one is a bit harder to take in.